File: dotNetTips.Utility/Extensions/ThreadExtensions.vb
' ***********************************************************************
' Assembly         : dotNetTips.Utility
' Author           : David McCarter
' Created          : 07-27-2017
'
' Last Modified By : David McCarter
' Last Modified On : 07-27-2017
' ***********************************************************************
' <copyright file="ThreadExtensions.vb" company="McCarter Consulting - David McCarter">
'     David McCarter - dotNetTips.com © 2017
' </copyright>
' <summary></summary>
' ***********************************************************************
Imports System.Runtime.CompilerServices
Imports System.Threading
Imports System.Windows.Forms

''' <summary>
''' Class ThreadExtensions.
''' </summary>
Public Module ThreadExtensions
    ''' <summary>
    ''' Waits the until the given date/ time.
    ''' </summary>
    ''' <param name="currentThread">The current thread.</param>
    ''' <param name="stopOn">The stop on.</param>
    <Extension>
    Public Sub WaitUntil(currentThread As Thread, stopOn As DateTime)
        Do Until DateTime.UtcNow >= stopOn.ToUniversalTime
            Application.DoEvents()
        Loop
    End Sub

    ''' <summary>
    ''' Waits for period of time.
    ''' </summary>
    ''' <param name="currentThread">The current thread.</param>
    ''' <param name="waitFor">The wait for.</param>
    <Extension>
    Public Sub WaitFor(currentThread As Thread, waitFor As TimeSpan)
        Dim stopOn = DateTime.UtcNow.Add(waitFor)
        currentThread.WaitUntil(stopOn)
    End Sub
End Module

File: dotNetTips.Utility.Standard.Extensions/ThreadExtensions.cs
// ***********************************************************************
// Assembly         : dotNetTips.Utility.Standard.Extensions
// Author           : David McCarter
// Created          : 09-15-2017
//
// Last Modified By : David McCarter
// Last Modified On : 09-18-2017
// ***********************************************************************
// <copyright file="ObjectExtensions.cs" company="dotNetTips.com - David McCarter">
//     dotNetTips.com - David McCarter
// </copyright>
// <summary></summary>
// ***********************************************************************
using System;
using System.Linq;
using System.Threading;

namespace dotNetTips.Utility.Standard.Extensions
{
    /// <summary>
    /// Class ThreadExtensions.
    /// </summary>
    public static class ThreadExtensions
    {
        /// <summary>
        /// Waits for a certian amount of time. Does not use timer (no need to call Dispose).
        /// </summary>
        /// <param name="thread">The thread.</param>
        /// <param name="interval">The interval.</param>
        public static void WaitUntil(this Thread thread, TimeSpan interval)
        {
            var stopAt = DateTime.Now.Add(interval);

            do
            {
                Thread.SpinWait(10);
            }
            while (thread.IsAlive && DateTime.Now < stopAt);

        }
    }
}